File: test/language/union_types/python/EmptyUnionTest.py
import unittest
import zserio

from testutils import getZserioApi

class EmptyUnionTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "union_types.zs").empty_union

    def testEmptyConstructor(self):
        emptyUnion = self.api.EmptyUnion()
        self.assertEqual(0, emptyUnion.bitsizeof())

    def testFromReader(self):
        reader = zserio.BitStreamReader([])
        emptyUnion = self.api.EmptyUnion.from_reader(reader)
        self.assertEqual(0, emptyUnion.bitsizeof())

    def testEq(self):
        emptyUnion1 = self.api.EmptyUnion()
        emptyUnion2 = self.api.EmptyUnion()
        self.assertTrue(emptyUnion1 == emptyUnion2)

    def testHash(self):
        emptyUnion1 = self.api.EmptyUnion()
        emptyUnion2 = self.api.EmptyUnion()
        self.assertEqual(hash(emptyUnion1), hash(emptyUnion2))

        # use hardcoded values to check that the hash code is stable
        # using __hash__ to prevent 32-bit Python hash() truncation
        self.assertEqual(850, emptyUnion1.__hash__())

    def testChoiceTag(self):
        emptyUnion = self.api.EmptyUnion()
        self.assertEqual(self.api.EmptyUnion.UNDEFINED_CHOICE, emptyUnion.choice_tag)

    def testBitSizeOf(self):
        bitPosition = 1
        emptyUnion = self.api.EmptyUnion()
        self.assertEqual(0, emptyUnion.bitsizeof(bitPosition))

    def testInitializeOffsets(self):
        bitPosition = 1
        emptyUnion = self.api.EmptyUnion()
        self.assertEqual(bitPosition, emptyUnion.initialize_offsets(bitPosition))

    def testRead(self):
        emptyUnion = self.api.EmptyUnion()
        reader = zserio.BitStreamReader([])
        emptyUnion.read(reader)
        self.assertEqual(0, emptyUnion.bitsizeof())

    def testWrite(self):
        writer = zserio.BitStreamWriter()
        emptyUnion = self.api.EmptyUnion()
        emptyUnion.write(writer)
        byteArray = writer.byte_array
        self.assertEqual(0, len(byteArray))
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readEmptyUnion = self.api.EmptyUnion.from_reader(reader)
        self.assertEqual(emptyUnion, readEmptyUnion)

File: test/language/union_types/cpp/EmptyUnionTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "union_types/empty_union/EmptyUnion.h"

namespace union_types
{
namespace empty_union
{

TEST(EmptyUnionTest, emptyConstructor)
{
    EmptyUnion emptyUnion;
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnion.choiceTag());
    ASSERT_EQ(0, emptyUnion.bitSizeOf());
}

TEST(EmptyUnionTest, bitStreamReaderConstructor)
{
    zserio::BitStreamReader reader(nullptr, 0);

    EmptyUnion emptyUnion(reader);
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnion.choiceTag());
    ASSERT_EQ(0, emptyUnion.bitSizeOf());
}

TEST(EmptyUnionTest, copyConstructor)
{
    EmptyUnion emptyUnion;
    EmptyUnion emptyUnionCopy(emptyUnion);
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnionCopy.choiceTag());
    ASSERT_EQ(0, emptyUnionCopy.bitSizeOf());
}

TEST(EmptyUnionTest, assignmentOperator)
{
    EmptyUnion emptyUnion;
    EmptyUnion emptyUnionCopy;
    emptyUnionCopy = emptyUnion;
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnionCopy.choiceTag());
    ASSERT_EQ(0, emptyUnionCopy.bitSizeOf());
}

TEST(EmptyUnionTest, moveConstructor)
{
    EmptyUnion emptyUnion;
    EmptyUnion emptyUnionMoved(std::move(emptyUnion));
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnionMoved.choiceTag());
    ASSERT_EQ(0, emptyUnionMoved.bitSizeOf());
}

TEST(EmptyUnionTest, moveAssignmentOperator)
{
    EmptyUnion emptyUnion;
    EmptyUnion emptyUnionMoved;
    emptyUnionMoved = std::move(emptyUnion);
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnionMoved.choiceTag());
    ASSERT_EQ(0, emptyUnionMoved.bitSizeOf());
}

TEST(EmptyUnionTest, propagateAllocatorCopyConstructor)
{
    EmptyUnion emptyUnion;
    EmptyUnion emptyUnionCopy(zserio::PropagateAllocator, emptyUnion, EmptyUnion::allocator_type());
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnionCopy.choiceTag());
    ASSERT_EQ(0, emptyUnionCopy.bitSizeOf());
}

TEST(EmptyUnionTest, choiceTag)
{
    EmptyUnion emptyUnion;
    ASSERT_EQ(EmptyUnion::UNDEFINED_CHOICE, emptyUnion.choiceTag());
}

TEST(EmptyUnionTest, bitSizeOf)
{
    EmptyUnion emptyUnion;
    const size_t bitPosition = 1;
    ASSERT_EQ(0, emptyUnion.bitSizeOf(bitPosition));
}

TEST(EmptyUnionTest, initializeOffsets)
{
    const size_t bitPosition = 1;
    EmptyUnion emptyUnion;
    ASSERT_EQ(bitPosition, emptyUnion.initializeOffsets(bitPosition));
}

TEST(EmptyUnionTest, operatorEquality)
{
    EmptyUnion emptyUnion1;
    EmptyUnion emptyUnion2;
    ASSERT_TRUE(emptyUnion1 == emptyUnion2);
}

TEST(EmptyUnionTest, hashCode)
{
    EmptyUnion emptyUnion1;
    EmptyUnion emptyUnion2;
    ASSERT_EQ(emptyUnion1.hashCode(), emptyUnion2.hashCode());
}

TEST(EmptyUnionTest, write)
{
    EmptyUnion emptyUnion;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    emptyUnion.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    EmptyUnion readEmptyUnion(reader);
    ASSERT_TRUE(emptyUnion == readEmptyUnion);
}

} // namespace empty_union
} // namespace union_types
